File: apps/api/src/auth/cookie.service.ts
import { Injectable } from '@nestjs/common';
import { CookieOptions, Request, Response } from 'express';
import { randomBytes } from 'crypto';
import { getConfig, parseSameSite } from '../config/env';

export const CSRF_HEADER_NAME = 'X-CSRF-TOKEN';

/**
 * Port of Procraft.Infrastructure.Auth.CookieService and
 * Procraft.Api.Extensions.CookieExtensions: identical cookie names, flags,
 * lifetimes and CSRF token format (uppercase hex, 32 chars).
 */
@Injectable()
export class CookieService {
  appendAccessToken(res: Response, jwt: string): void {
    if (!jwt || jwt.trim() === '') {
      return;
    }

    const config = getConfig();
    const expires = new Date(Date.now() + config.jwt.accessTokenMinutes * 60_000);
    res.cookie(config.cookies.accessCookieName, jwt, this.buildOptions(expires));
  }

  appendRefreshToken(res: Response, plaintextRefreshToken: string): void {
    if (!plaintextRefreshToken || plaintextRefreshToken.trim() === '') {
      return;
    }

    const config = getConfig();
    const expires = new Date(Date.now() + config.jwt.refreshTokenDays * 86_400_000);
    res.cookie(config.cookies.refreshCookieName, plaintextRefreshToken, this.buildOptions(expires));
  }

  getPlainRefreshToken(req: Request): string | null {
    const config = getConfig();
    const value = (req.cookies ?? {})[config.cookies.refreshCookieName];
    return typeof value === 'string' ? value : null;
  }

  clearAuthCookies(res: Response): void {
    const config = getConfig();
    const options = this.buildOptions();
    res.clearCookie(config.cookies.accessCookieName, options);
    res.clearCookie(config.cookies.refreshCookieName, options);
  }

  /** Non-HttpOnly double-submit CSRF cookie (hex of 16 random bytes, uppercase). */
  issueCsrfTokenCookie(res: Response): void {
    const config = getConfig();
    const token = randomBytes(16).toString('hex').toUpperCase();

    const options: CookieOptions = {
      httpOnly: false,
      secure: config.cookies.secure,
      sameSite: parseSameSite(config.cookies.sameSite),
      path: '/',
    };

    if (config.cookies.csrfCookieDomain && config.cookies.csrfCookieDomain.trim() !== '') {
      options.domain = config.cookies.csrfCookieDomain;
    }

    res.cookie(config.cookies.csrfCookieName, token, options);
  }

  private buildOptions(expires?: Date): CookieOptions {
    const config = getConfig();

    const options: CookieOptions = {
      httpOnly: true,
      secure: config.cookies.secure,
      sameSite: parseSameSite(config.cookies.sameSite),
      path: '/',
    };

    if (expires) {
      options.expires = expires;
      options.maxAge = expires.getTime() - Date.now();
    }

    return options;
  }
}

File: apps/api-dotnet/src/Procraft.Infrastructure/Auth/CookieService.cs
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Options;
using Procraft.Application.Common.Configuration;
using Procraft.Application.Common.Interfaces;
using Procraft.Application.Common.Security;

namespace Procraft.Infrastructure.Auth;

public sealed class CookieService : ICookieService
{
    private readonly IHttpContextAccessor _http;
    private readonly AuthCookieOptions _options;
    private readonly JwtOptions _jwt;

    public CookieService(
        IHttpContextAccessor http,
        IOptions<AuthCookieOptions> options,
        IOptions<JwtOptions> jwt)
    {
        _http = http;
        _options = options.Value;
        _jwt = jwt.Value;
    }

    private HttpContext Context =>
        _http.HttpContext ?? throw new InvalidOperationException("HttpContext is not available.");

    public void AppendAccessToken(string jwt)
    {
        if (string.IsNullOrWhiteSpace(jwt))
        {
            return;
        }

        Context.Response.Cookies.Append(
            _options.AccessCookieName,
            jwt,
            BuildOptions(DateTimeOffset.UtcNow.AddMinutes(_jwt.AccessTokenMinutes)));
    }

    public void AppendRefreshToken(string plaintextRefreshToken)
    {
        if (string.IsNullOrWhiteSpace(plaintextRefreshToken))
        {
            return;
        }

        Context.Response.Cookies.Append(
            _options.RefreshCookieName,
            plaintextRefreshToken,
            BuildOptions(DateTimeOffset.UtcNow.AddDays(_jwt.RefreshTokenDays)));
    }

    public string? GetPlainRefreshToken() =>
        Context.Request.Cookies.TryGetValue(_options.RefreshCookieName, out var raw) ? raw : null;

    public void ClearAuthCookies()
    {
        var opts = BuildOptions();
        Context.Response.Cookies.Delete(_options.AccessCookieName, opts);
        Context.Response.Cookies.Delete(_options.RefreshCookieName, opts);
    }

    private CookieOptions BuildOptions(DateTimeOffset? expires = null)
    {
        var sameSite = SameSiteMode.Lax;
        if (Enum.TryParse<SameSiteMode>(_options.SameSite, true, out var parsed))
        {
            sameSite = parsed;
        }

        return new CookieOptions
        {
            HttpOnly = true,
            Secure = _options.Secure,
            SameSite = sameSite,
            Path = "/",
            Expires = expires,
            MaxAge = expires.HasValue ? expires.Value - DateTimeOffset.UtcNow : null,
        };
    }
}
